spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlignMergeIntoTable.scala

```diff
@@ -0,0 +1,91 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.plans.logical.{Assignment, DeleteAction, InsertAction, LogicalPlan, MergeIntoTable, UpdateAction}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.internal.SQLConf
+
+case class AlignMergeIntoTable(conf: SQLConf) extends Rule[LogicalPlan] with AssignmentAlignmentSupport {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case m: MergeIntoTable if m.resolved =>
+      val alignedMatchedActions = m.matchedActions.map {
+        case u @ UpdateAction(_, assignments) =>
+          u.copy(assignments = alignAssignments(m.targetTable, assignments))
+        case d: DeleteAction => d
+      }
+
+      val alignedNotMatchedActions = m.notMatchedActions.map {
+        case i @ InsertAction(_, assignments) =>
+          // check no nested columns are present
+          val namePartsSeq = assignments.map(_.key).map(getNameParts)
+          namePartsSeq.foreach { nameParts =>
+            if (nameParts.size > 1) {
+              throw new AnalysisException(
+                "Nested fields are not supported inside INSERT clauses of MERGE operations: " +
+                s"${nameParts.mkString("`", "`.`", "`")}")
+            }
+          }
+
+          val colNames = namePartsSeq.map(_.head)
+
+          // check there are no duplicates
+          val duplicateColNames = colNames.groupBy(identity).collect {
+            case (name, matchingNames) if matchingNames.size > 1 => name
+          }
+
+          if (duplicateColNames.nonEmpty) {
+            throw new AnalysisException(
+              "Duplicate column names inside INSERT clause: " +
+              duplicateColNames.mkString("[", ",", "]"))
+          }
+
+          // reorder assignments by the target table column order
+          i.copy(assignments = alignInsertActionAssignments(m.targetTable, colNames, assignments))
+        case _ =>
+          throw new AnalysisException("Not matched actions can only contain INSERT")
+      }
+
+      m.copy(matchedActions = alignedMatchedActions, notMatchedActions = alignedNotMatchedActions)
+  }
+
+  private def alignInsertActionAssignments(
+      targetTable: LogicalPlan,
+      insertCols: Seq[String],
+      assignments: Seq[Assignment]): Seq[Assignment] = {
+
+    val resolver = conf.resolver
+    targetTable.output.map { targetAttr =>
+      val assignment = assignments.find(a => resolver(targetAttr.name, getNameParts(a.key).head))
```

Comment:
This could avoid calling `getNameParts` by creating a map out of `insertCols` and `assignments`:

```
val assignmentMap = insertCols.zip(assigments).toMap
```

That would be better since this is going to call `getNameParts` for every assignment until a matching one is found.

Reply:
Good point.

Reply:
I think a simple zip won't work since the assignments haven't been aligned yet. Plus, we have to use the resolver for equality so cannot have O(1) lookups.

That said, we can avoid the main problem of constantly calling `getNameParts`. Let me do that.

Reply:
I thought insertCols was derived from assignments like `assigments.map(getNameParts(_.key).head)`, so we know that they are equivalent. You're right about the resolver, though.

Reply:
Yeah, you are right. I can simplify this a bit further.

Reply:
Done.